### PR TITLE
fix: sharky waitgroup handling

### PR DESCRIPTION
**Pull request by janos**
### Checklist

- [ ] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md)
- [ ] My change requires a documentation update and I have done it
- [ ] I have added tests to cover my changes.

### Description
This PR is an attempt to use wait group in more predictable way in order to fix the issue with closing the sharky.

Testing these changes on integration would be needed as unit tests are injecting wait group which removes a bit of reliability from them.  

#### Motivation and context (Optional)
Make shutdown process reliable.

### Related Issue (Optional)
https://github.com/ethersphere/bee/issues/2840

### Screenshots (if appropriate):

---
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/2841)

### CODE REVIEW

**Review by mrekucci — APPROVED** [review]
Thanks!

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2841#-:-MxieZ0v8Fcov-udxPio:b-ucfknh)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @aloknerurkar and @ldeffenb)

**Review by ldeffenb — APPROVED** [review]
LGTM but I admit to not completely understanding the flow through sharky yet.  I'll get this merged in with my testnet rc2 build and give it a whirl.

**Comment by janos** [comment]
> LGTM but I admit to not completely understanding the flow through sharky yet. I'll get this merged in with my testnet rc2 build and give it a whirl.

@ldeffenb Great, thanks. That would be the most important test, currently. I have removed the commented code that I missed. And I suggest to merge this pr after we get the feedback from tests from you on your infrastructure.

File: pkg/sharky/shard.go
Comment (janos) [inline pkg/sharky/shard.go:134]: There may be a possibility to have slots not read from this channel. I think that there is still need some improvement around insurance that all slots are pushed before quit.

Comment (janos) [reply pkg/sharky/shard.go]: I have added another wg for the "limbo" free slots writes to wait for them after the quit is closed. That way the original waitgroup is easier to follow but also ensures that no free slots are dropped.

**Comment by janos** [comment]
@ldeffenb I have made additional and crucial improvements. May I ask for a test on the latest state of this branch?

**Comment by ldeffenb** [comment]
Well, there is no joy in Mudville, the wait-group change didn't work.  The shutdown still hung.  My trace logs in node.go and localstore.go show:
```
time="2022-03-09T09:59:30-05:00" level=info msg="Shutdown stateStoreCloser"
time="2022-03-09T09:59:30-05:00" level=info msg="Shutdown localstoreCloser"
time="2022-03-09T09:59:30-05:00" level=info msg="localstore.Close close(db.close)"
time="2022-03-09T09:59:30-05:00" level=info msg="localstore.Close cancel()"
time="2022-03-09T09:59:30-05:00" level=info msg="localstore.Close updateGCWG.Wait()"
time="2022-03-09T09:59:30-05:00" level=info msg="localstore.Close collectGarbageWorkderDone"
time="2022-03-09T09:59:30-05:00" level=info msg="localstore.Close reserveEvictionWorkerDone"
time="2022-03-09T09:59:30-05:00" level=info msg="localstore.Close close(done)"
time="2022-03-09T09:59:30-05:00" level=info msg="localstore.Close close(done) done"
time="2022-03-09T09:59:30-05:00" level=info msg="localstore.Close select <-done"
time="2022-03-09T09:59:30-05:00" level=info msg="localstore.Close db.sharky.Close()"
```

And on a semi-related note, since the localstore is hung on shutdown, apparently the stamp issuer is still getting accessed by someone because I also got the following logs immediately after the previous set, but 1 second later.
```
time="2022-03-09T09:59:31-05:00" level=debug msg="get stamp issuer: check batch: leveldb: closed"
time="2022-03-09T09:59:31-05:00" level=error msg="get stamp issuer: check batch"
```

The debug/pprof/goroutine output before the Control-C is at: 
https://gateway.ethswarm.org/access/713d3d0c32b7a4a203a9d709d19e6d994f26213945372cd1042641dfcf32c16c

The goroutines after the Control-C and the subsequent hang is at:
https://gateway.ethswarm.org/access/0648f6384d44e80b0e611d77b8e83e9bffcfda96cd3b32a714da692fad5a85c7

Note: Your line numbers in node.go and localstore.go will not match because I have a dirty copy with additional shutdown trace logs.  The sharky files are straight out of the sharky-wg-handling branch.  If you want a copy of my dirty logs, you can get them from:
node.go: https://gateway.ethswarm.org/access/d28f01866ccfae5c03022009a5831dea4a62b68d481a5e9f926397c1a5b2fdc8
localstore.go: https://gateway.ethswarm.org/access/e14742fe4330cc4fea7b2e6ad2e647827c2dba082d83a06324871bbe7f588d52

**Comment by ldeffenb** [comment]
> @ldeffenb I have made additional and crucial improvements. May I ask for a test on the latest state of this branch?

By all means.  I'll get the latest pulled in with my dirty shutdown logging and give it a go.  Watch this space over the next few hours.

**Review by aloknerurkar — APPROVED** [review]

**Review by mrekucci — APPROVED** [review]
> Reviewed 1 of 1 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2841)* status: all files reviewed, 2 unresolved discussions (waiting on @janos and @ldeffenb)

**Comment by ldeffenb** [comment]
Nope, db1bc96f still hung on shutdown with a single thread left in the shard process routine.  The goroutines debug can be pulled from:

https://gateway.ethswarm.org/access/dd59221ad24dbc3bb50837c11f41e39be1261d3cfe7e737a6c8a9a74e6eb846e

Let me know when to try again.  I also have the verbosity 5 logs of this node, just ask me for bee1-150rc2-wg2-0.out

Note the localstore and node line numbers will not match your's, but are the same as the dirty source files I linked earlier.

**Comment by janos** [comment]
@ldeffenb your testing and information that you sent are very useful and valuable. The same issue was present for the reads as well, that I did not address, previously. Another update would remove deadlock on read operation and I hope that that is all. I have to ask for another test on the latest state of this branch https://github.com/ethersphere/bee/pull/2841/commits/b7a2c9eaac4b750f9e451ae9409c91bdac525ccc.

**Comment by ldeffenb** [comment]
Still testing.  My depth keeps crashing to zero which completely changes what the node is doing.  But in the meantime, you can ignore my earlier comment about stamps accessing the localstore after it is closed.  I got the following logs on my last shutdown which indicates that it's my monBee watching the node that is doing a stamp query coming in via the (dirty) still-enabled debug API that is finding the localstore closed.
```
time="2022-03-09T15:30:04-05:00" level=info msg="localstore.Close db.sharky.Close()"
time="2022-03-09T15:30:04-05:00" level=debug msg="get stamp issuer: check batch: leveldb: closed"
time="2022-03-09T15:30:04-05:00" level=error msg="get stamp issuer: check batch"
time="2022-03-09T15:30:04-05:00" level=info msg="debug api access" duration=0.0009226 ip=192.168.10.177 method=GET proto=HTTP/1.1 size=48 status=500 uri=/stamps/c749612363aa9f0345041e24b54e6d177285148ac78356a6b33e468d6b418995
time="2022-03-09T15:30:04-05:00" level=info msg="localstore.Close db.shed.Close()"
```

**Comment by janos** [comment]
Thank you for the update and clarification @ldeffenb.

**Comment by ldeffenb** [comment]
After looking closely at my logs when the depth crashes, I found hints to check my local node's host time synchronization.  It was 15 seconds slow.  Meinbert NTP has been added to keep the time correctly synced (gotta love/hate Windows).  Starting a new test now that the machine knows better what time it is.

**Review by ldeffenb — APPROVED** [review]
b7a2c9ea looks like it has fixed my pusher hanging and also shuts down cleanly.  I restarted once or twice without issue and my pusher has been running steady for over 6 hours now.   I just stopped it again cleanly while it still has chunks pending push to test after this PR is merged.

**Comment by janos** [comment]
Wonderful, thanks @ldeffenb for helping us!

**Pull request merged**
